torchao/dtypes/affine_quantized_tensor.py

```diff
@@ -329,37 +339,47 @@ def from_hp_to_intx(
             )
             data = data.to(target_dtype)
         else:
-            scale, zero_point = choose_qparams_affine(
-                input_float,
-                mapping_type,
-                block_size,
-                target_dtype,
-                quant_min,
-                quant_max,
-                eps,
-                scale_dtype,
-                zero_point_dtype,
-                preserve_zero,
-                zero_point_domain,
-            )
-            # choose_qparams_affine is a custom op that does support returning optional Tensors. We thus set the zero_point to None if its domain is None
-            if zero_point_domain is None:
-                zero_point = None
-            data = quantize_affine(
-                input_float,
-                block_size,
-                scale,
-                zero_point,
-                target_dtype,
-                quant_min,
-                quant_max,
-                zero_point_domain,
-            )
-            # Note: output will be uint8 tensor for sub byte tensors for now
+            if isinstance(_layout, MarlinQQQLayout):
+                nbits = int(math.log2(quant_max - quant_min + 1))
+                group_size = max(block_size)
+                data, s_group, s_channel, _ = choose_qparams_and_quantize_affine_qqq(
+                    input_float, nbits, group_size
+                )
+            else:
+                scale, zero_point = choose_qparams_affine(
+                    input_float,
+                    mapping_type,
+                    block_size,
+                    target_dtype,
+                    quant_min,
+                    quant_max,
+                    eps,
+                    scale_dtype,
+                    zero_point_dtype,
+                    preserve_zero,
+                    zero_point_domain,
+                )
+                # choose_qparams_affine is a custom op that does support returning optional Tensors. We thus set the zero_point to None if its domain is None
+                if zero_point_domain is None:
+                    zero_point = None
+                data = quantize_affine(
+                    input_float,
+                    block_size,
+                    scale,
+                    zero_point,
+                    target_dtype,
+                    quant_min,
+                    quant_max,
+                    zero_point_domain,
+                )
+                # Note: output will be uint8 tensor for sub byte tensors for now
 
         data = _layout.post_process(data)
         tensor_impl_ctr = get_tensor_impl_constructor(type(_layout))
-        tensor_impl = tensor_impl_ctr(data, scale, zero_point, _layout)
+        if isinstance(_layout, MarlinQQQLayout):
+            tensor_impl = tensor_impl_ctr(data, s_group, s_channel, _layout)
```

Comment:
yeah it feels like it's better to create a new tensor subclass, we defined affine quantization being: https://github.com/pytorch/ao/blob/0e854ecbf75350d781fa60140ab1d3180a0404db/torchao/dtypes/affine_quantized_tensor.py#L136-L137 does it still hold for MarlinQQQ?

Reply:
yeah that sounds good to me

Reply:
I have modified, please check it.